### PR TITLE
aggregate stats function

### PR DESCRIPTION
**Pull request by guga31bb**

### CODE REVIEW

**Comment by guga31bb** [comment]
Not ready but I sketched what I think it would look like. The latter two replicate our new work in the .rmd in the other branch. 

``` r
pbp <- load_pbp(2020)
> get_player_stats(pbp)
# A tibble: 5,434 x 19
   player_id player_name season  week   cmp   att passing_yards pass_tds  ints fumble_lost_sack carries
   <chr>     <chr>        <int> <int> <int> <int>         <dbl>    <int> <dbl>            <dbl>   <int>
 1 00-00195~ T.Brady       2020     1    23    36           239        2     2                0       3
 2 00-00195~ T.Brady       2020     2    23    35           217        1     1                0       1
 3 00-00195~ T.Brady       2020     3    25    38           297        3     0                0       5
 4 00-00195~ T.Brady       2020     4    30    46           369        5     1                0       3
 5 00-00195~ T.Brady       2020     5    25    41           253        1     0                1       3
 6 00-00195~ T.Brady       2020     6    17    27           166        2     0                0       0
 7 00-00195~ T.Brady       2020     7    33    45           369        4     0                0       1
 8 00-00195~ T.Brady       2020     8    28    40           279        2     0                1       1
 9 00-00195~ T.Brady       2020     9    22    38           209        0     3                0       0
10 00-00195~ T.Brady       2020    10    28    39           341        3     0                1       2
# ... with 5,424 more rows, and 8 more variables: rushing_yards <dbl>, tds_rush <int>,
#   fumble_lost_rush <dbl>, rec <int>, tgt <int>, receiving_yards <dbl>, tds_rec <int>,
#   fumble_lost_rec <dbl>

> get_player_stats(pbp %>% dplyr::filter(week <= 17), weekly = FALSE) %>% dplyr::arrange(-passing_yards)
# A tibble: 642 x 17
# Groups:   player_id [628]
   player_id player_name   cmp   att passing_yards pass_tds  ints fumble_lost_sack carries
   <chr>     <chr>       <int> <int>         <dbl>    <int> <dbl>            <dbl>   <int>
 1 00-00335~ D.Watson      382   544          4823       33     7                5      90
 2 00-00338~ P.Mahomes     390   588          4740       38     6                7      62
 3 00-00195~ T.Brady       401   610          4633       40    12                3      30
 4 00-00261~ M.Ryan        407   626          4581       26    11                4      29
 5 00-00348~ J.Allen       396   572          4544       37    10                4     102
 6 00-00363~ J.Herbert     396   595          4336       31    10                3      55
 7 00-00234~ A.Rodgers     372   526          4299       48     5                3      38
 8 00-00296~ K.Cousins     349   516          4265       35    13                6      32
 9 00-00292~ R.Wilson      384   558          4212       40    13                3      83
10 00-00229~ P.Rivers      369   543          4169       24    11                3      18
# ... with 632 more rows, and 8 more variables: rushing_yards <dbl>, tds_rush <int>,
#   fumble_lost_rush <dbl>, rec <int>, tgt <int>, receiving_yards <dbl>, tds_rec <int>,
#   fumble_lost_rec <dbl>

> get_player_stats(pbp %>% dplyr::filter(week <= 17), weekly = FALSE) %>% dplyr::arrange(-rushing_yards) %>% dplyr::select(player_name, carries, rushing_yards, tds_rush)
Adding missing grouping variables: `player_id`
# A tibble: 642 x 5
# Groups:   player_id [628]
   player_id  player_name  carries rushing_yards tds_rush
   <chr>      <chr>          <int>         <dbl>    <int>
 1 00-0032764 D.Henry          378          2027       17
 2 00-0033893 D.Cook           312          1557       16
 3 00-0036223 J.Taylor         232          1169       11
 4 00-0033293 A.Jones          201          1104        9
 5 00-0035685 D.Montgomery     247          1070        8
 6 00-0035831 J.Robinson       240          1070        7
 7 00-0034791 N.Chubb          190          1067       12
 8 00-0035700 J.Jacobs         273          1065       12
 9 00-0034796 L.Jackson        159          1005        7
10 00-0032144 M.Gordon         215           986        9
```

**Comment by mrcaseb** [comment]
Looks pretty solid to me! 

The last thing we are missing now are plays with multiple laterals. The pbp data will only host the last player with lateral yards (either rushing or receiving). So in plays where multiple players recorded lateral yards some of them will not be available in the pbp data. That's why I have [this list](https://github.com/mrcaseb/nfl-data/blob/master/data/lateral_yards/multiple_lateral_yards.csv). Passing yards are not affected as they include those yards. 
So we just have to join the list to rushing and receiving yards and should be perfect.

``` r
library(dplyr, warn.conflicts = FALSE)
library(nflfastR)
pbp <- load_pbp(2020)

pbp %>% 
  filter(week <= 17) %>% 
  get_player_stats(weekly = TRUE) %>% 
  select(season, week, player_name, rec, receiving_yards, tds_rec) %>%
  filter(player_name == "B.Cooks")
#> # A tibble: 15 x 6
#>    season  week player_name   rec receiving_yards tds_rec
#>     <int> <int> <chr>       <int>           <dbl>   <int>
#>  1   2020     1 B.Cooks         2              20       0
#>  2   2020     2 B.Cooks         5              95       0
#>  3   2020     3 B.Cooks         3              23       0
#>  4   2020     4 B.Cooks         0               0       0
#>  5   2020     5 B.Cooks         8             161       1
#>  6   2020     6 B.Cooks         9              68       1
#>  7   2020     7 B.Cooks         7              60       0
#>  8   2020     9 B.Cooks         3              83       1
#>  9   2020    10 B.Cooks         6              39       0
#> 10   2020    11 B.Cooks         4              85       0
#> 11   2020    12 B.Cooks         5              85       0
#> 12   2020    13 B.Cooks         5              65       0
#> 13   2020    15 B.Cooks         6              59       0
#> 14   2020    16 B.Cooks         7             134       1
#> 15   2020    17 B.Cooks        11             166       2
```

<sup>Created on 2021-02-16 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

grafik

**Comment by mrcaseb** [comment]
Added the plays with multiple laterals

``` r
library(dplyr, warn.conflicts = FALSE)
library(nflfastR)
pbp <- load_pbp(2020)

pbp %>% 
  filter(week <= 17) %>% 
  get_player_stats(weekly = TRUE) %>% 
  select(season, week, player_name, rec, receiving_yards, tds_rec) %>%
  filter(player_name == "B.Cooks")
#> # A tibble: 15 x 6
#>    season  week player_name   rec receiving_yards tds_rec
#>     <int> <int> <chr>       <int>           <dbl>   <int>
#>  1   2020     1 B.Cooks         2              20       0
#>  2   2020     2 B.Cooks         5              95       0
#>  3   2020     3 B.Cooks         3              23       0
#>  4   2020     4 B.Cooks         0               0       0
#>  5   2020     5 B.Cooks         8             161       1
#>  6   2020     6 B.Cooks         9              68       1
#>  7   2020     7 B.Cooks         7              60       0
#>  8   2020     9 B.Cooks         3              83       1
#>  9   2020    10 B.Cooks         6              39       0
#> 10   2020    11 B.Cooks         4              85       0
#> 11   2020    12 B.Cooks         5              85       0
#> 12   2020    13 B.Cooks         5              65       0
#> 13   2020    15 B.Cooks         6              59       0
#> 14   2020    16 B.Cooks         7             141       1
#> 15   2020    17 B.Cooks        11             166       2
```

<sup>Created on 2021-02-16 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by mrcaseb** [comment]
For comparison here's what happens if the IDs are not encoded. 
``` r
library(dplyr, warn.conflicts = FALSE)
library(nflfastR)

pbp <- fast_scraper_schedules(2020) %>% 
  filter(week == 16) %>%
  fast_scraper() 
  
pbp %>% 
  get_player_stats(weekly = TRUE) %>% 
  select(season, week, player_name, rec, receiving_yards, tds_rec) %>%
  filter(player_name == "B.Cooks")
#> # A tibble: 1 x 6
#>   season  week player_name   rec receiving_yards tds_rec
#>    <int> <int> <chr>       <int>           <dbl>   <int>
#> 1   2020    16 B.Cooks         7             134       1

pbp %>% 
  decode_player_ids() %>% 
  get_player_stats(weekly = TRUE) %>% 
  select(season, week, player_name, rec, receiving_yards, tds_rec) %>%
  filter(player_name == "B.Cooks")
#> v Decoding of player ids completed
#> # A tibble: 1 x 6
#>   season  week player_name   rec receiving_yards tds_rec
#>    <int> <int> <chr>       <int>           <dbl>   <int>
#> 1   2020    16 B.Cooks         7             141       1
```

<sup>Created on 2021-02-16 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by guga31bb** [comment]
> This will only work if the IDs in the param `pbp` are decoded. Shall we do this internally in the function?

Yes I think that makes sense

File: R/aggregate_game_stats.R [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  if (isFALSE(weekly)) {
+    player_df <- player_df %>%
+      dplyr::group_by(.data$player_id, .data$player_name) %>%
+      dplyr::summarise(
+        completions = sum(.data$completions),
+        attempts = sum(.data$attempts),
+        passing_yards = sum(.data$passing_yards),
+        passing_tds = sum(.data$passing_tds),
+        interceptions = sum(.data$interceptions),
+        passing_air_yards = sum(.data$passing_yards),
```
Comment (mrcaseb) [inline R/aggregate_game_stats.R:222]: This doesn't use air yards

**Comment by mrcaseb** [comment]
I like the air yards idea. Shall we add yac as well?

If we do so we can't use the yac column. Instead compute passing_yards - air_yards

**Comment by mrcaseb** [comment]
> I like the air yards idea. Shall we add yac as well?
> 
> If we do so we can't use the yac column. Instead compute passing_yards - air_yards

How to compute yac:

- `passing_yac` = `passing_yards` - `passing_aur_yards`

- `receiving_yac` = `sum(yards_after_catch)` + `lateral_yards`

**Comment by guga31bb** [comment]
Ok so all that's left is figuring out the YAC lateral thing

**Comment by guga31bb** [comment]
Oh and decoding the IDs

**Comment by mrcaseb** [comment]
I looked into the YAC stuff. I don't think this is easy to fix (if at all). The below code compares nflfastR, ESPN and NGS.
More or less big differences across all of them. So I think we just keep it as it is now and I have added some caveat text to the yac description.

``` r
library(tidyverse)
library(nflfastR)
library(espnscrapeR)

df <- nflfastR:::load_ngs(2020, "receiving")
pbp <- load_pbp(2020, TRUE) %>% filter(season_type == "REG")

ngs <- df %>% 
  select(season:week, player_short_name, receptions:avg_yac, player_gsis_id) %>% 
  arrange(-yards, receptions) %>% 
  filter(week == 0) %>%
  mutate(ngs_yac = (avg_yac * receptions) %>% round()) %>% 
  head(20)

espn <- scrape_espn_stats(season = 2020, stats = "receiving", season_type = "Regular") %>% 
  arrange(-rec_yards, receptions) %>% 
  head(20)
#> Scraping receiving stats from 2020 Regular season!

nflfastr <- get_player_stats(pbp)

comb <- nflfastr %>% 
  select(player_id:recent_team, receptions:receiving_yards_after_catch) %>% 
  filter(player_id %in% ngs$player_gsis_id) %>% 
  arrange(-receiving_yards, receptions) %>% 
  head(20) %>% 
  bind_cols(
    ngs %>% select(ngs_yac),
    espn %>% select(espn_yac = yards_after_catch) 
  )

comb %>% 
  ggplot(aes(x = seq_along(player_name))) +
  geom_point(aes(y = receiving_yards_after_catch, color = "nflfastR"), size = 3) +
  geom_point(aes(y = ngs_yac, color = "NGS"), size = 3) +
  geom_point(aes(y = espn_yac, color = "ESPN"), size = 3) +
  scale_x_continuous(breaks = seq_along(comb$player_name), labels = comb$player_name) +
  labs(y = "Yards After Catch") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5), axis.title.x = element_blank())
```



<sup>Created on 2021-02-17 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by mrcaseb** [comment]
Final comment on the yac thing. I think we can be confident in the numbers. Compared to PFR we are perfect (including Brandin Cooks who had lateral yac). A. J. Browns TDs differ, so this might be a last bug
``` r
library(tidyverse)
library(nflfastR)

df <- nflfastR:::load_ngs(2020, "receiving")
pbp <- load_pbp(2020, TRUE) %>% filter(season_type == "REG")

calculate_player_stats(pbp) %>% 
  select(player_name, receptions:receiving_tds, receiving_yards_after_catch) %>% 
  arrange(-receiving_yards, receptions) %>% 
  head(20)
#> # A tibble: 20 x 6
#>    player_name receptions targets receiving_yards receiving_tds receiving_yards~
#>    <chr>            <int>   <int>           <dbl>         <int>            <dbl>
#>  1 S.Diggs            127     166            1535             8              464
#>  2 T.Kelce            105     145            1416            11              587
#>  3 D.Hopkins          115     160            1407             6              534
#>  4 J.Jefferson         88     125            1400             7              453
#>  5 C.Ridley            90     143            1374             9              275
#>  6 D.Adams            115     149            1374            18              597
#>  7 D.Metcalf           83     129            1303            10              363
#>  8 T.Hill              87     135            1276            15              434
#>  9 A.Robinson         102     151            1250             6              319
#> 10 D.Waller           107     145            1196             9              572
#> 11 D.Moore             66     118            1193             4              382
#> 12 B.Cooks             81     119            1150             6              350
#> 13 T.McLaurin          87     134            1118             4              478
#> 14 A.Cooper            92     130            1114             5              413
#> 15 R.Anderson          95     136            1096             3              497
#> 16 A.Brown             70     106            1075            12              432
#> 17 T.Lockett          100     132            1054            10              327
#> 18 M.Evans             70     109            1006            13              265
#> 19 K.Allen            100     147             992             8              437
#> 20 C.Davis             65      92             984             5              291
```

grafik

**Comment by mrcaseb** [comment]
Checks will fail now until we merge #192 and update the data repo

**Comment by mrcaseb** [comment]
Some ideas on possible additions:
- [ ] 2 pt. conversions
- [ ] 1st downs
- [ ] avg. and total EPA
- [ ] actually with 2pt we have everything we need for fantasy points and ppr (like PFR does) so it's easy to just add them
- [ ] sacks
- [ ] aDOT, both passer and receiver

**Comment by guga31bb** [comment]
I would say yes to all these except the adot ones since those are just
combining 2 columns we made

**Comment by mrcaseb** [comment]
Added 2 pt conversions. Numbers can be compared [with this source](https://www.teamrankings.com/nfl/player-stat/scoring-two-point-conversions). I am confident this is correct. Overall 112 is an exact match and the leaders are the same as well

**Comment by mrcaseb** [comment]
Top 20 PPR points compared with https://fantasy.nfl.com/research/scoringleaders is an exact match
grafik

**Comment by guga31bb** [comment]
I didn't bother with average EPA since that's just total EPA / plays, but the rest are there now

**Pull request merged**